exec/java-exec/src/test/java/org/apache/drill/TestDynamicUDFSupport.java

```diff
@@ -0,0 +1,292 @@
+    assertEquals("Registry should contain one jar", registry.getJarList().size(), 1);
+    assertEquals(registry.getJar(0).getName(), jName);
+    test(String.format("drop function using jar '%s'", jName));
+  }
+
+  @Test
+  public void testDuplicatedJar() throws Exception {
+    copyJarsToStagingArea();
+    test(String.format("create function using jar '%s'", jName));
+    copyJarsToStagingArea();
+
+    testBuilder()
+        .sqlQuery(String.format("create function using jar '%s'", jName))
+        .unOrdered()
+        .baselineColumns("ok", "summary")
+        .baselineValues(false, String.format("Jar with %s name has been already registered", jName))
+        .go();
+
+    test(String.format("drop function using jar '%s'", jName));
+  }
+
+  @Test
+  public void testDuplicatedFunctions() throws Exception {
+    copyJarsToStagingArea();
+    test(String.format("create function using jar '%s'", jName));
+    copyJarsToStagingArea();
+
+    RemoteFunctionRegistry remoteFunctionRegistry = getDrillbitContext().getRemoteFunctionRegistry();
+    FileSystem fs = remoteFunctionRegistry.getFs();
+    String newJar = "DrillUDF-2.0.jar";
+    String newSource = JarUtil.getSourceName(newJar);
+    Path renamedBinary = new Path(remoteFunctionRegistry.getStagingArea(), newJar);
+    Path renamedSource = new Path(remoteFunctionRegistry.getStagingArea(), newSource);
+    fs.rename(new Path(remoteFunctionRegistry.getStagingArea(), jName), renamedBinary);
+    fs.rename(new Path(remoteFunctionRegistry.getStagingArea(), sName), renamedSource);
+
+    String summary = "Found duplicated function in %s - name: \"custom_lower\"\n" +
+        "major_type {\n" +
+        "  minor_type: VARCHAR\n" +
+        "  mode: REQUIRED\n" +
+        "}\n";
+
+    testBuilder()
+        .sqlQuery(String.format("create function using jar '%s'", newJar))
+        .unOrdered()
+        .baselineColumns("ok", "summary")
+        .baselineValues(false, String.format(summary, jName))
+        .go();
+
+    fs.delete(renamedBinary, false);
+    fs.delete(renamedSource, false);
+    test(String.format("drop function using jar '%s'", jName));
+  }
+
+  @Test
+  public void testLazyInit() throws Exception {
+    try {
+      test("select custom_lower(version) from sys.version");
+    } catch (UserRemoteException e){
+      assertThat(e.getMessage(), containsString("No match found for function signature custom_lower(<CHARACTER>)"));
+    }
+
+    copyJarsToStagingArea();
+
+    test(String.format("create function using jar '%s'", jName));
+    test("select custom_lower(version) from sys.version");
+
+    assertTrue("Binary should be present in local UDF area", new File(System.getenv("DRILL_UDF_DIR"), jName).exists());
+    assertTrue("Source should be present in local UDF area", new File(System.getenv("DRILL_UDF_DIR"), sName).exists());
+    test(String.format("drop function using jar '%s'", jName));
+  }
+
+  @Test
+  public void testDropFunction() throws Exception {
```

Comment:
Maybe a test of the trick case in which user A registers a UDF and starts a long-running query that uses the UDF. User B deregisters that same function. Does the query for user A continue to run, perhaps using a cached version of the function?

Then, user B registers a new UDF of the same name. Does the registration work? Should it? If it does, does it create a new copy of the function so that the old and new versions exist in memory? Now, user B runs a query. It should use the second UDF. But, user A's query should still be running using the first version.

This will work if the functions are loaded in separate class loaders, and the generate code is aware of which class loader to use for each function, so that it resolves the correct one. This, in turn, means that function lookup must be dynamic (by name, based on class loader), not implicit (which uses a single class loader tree.)

Such a test case might 1) prove the code is rock solid, or 2) suggest ways of enhancing the code, or 3) point out inherient limitations in the design that we must document (if the above use case causes failures and we accept those failures as expected behavior.)